blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc

```diff
@@ -0,0 +1,195 @@
+= Migrating OptaWeb Employee Rostering Backend Application to Spring Boot
+:page-interpolate: true
+:awestruct-author: jucui
+:awestruct-layout: blogPostBase
+:awestruct-tags: [employee rostering, coding, algorithm, production]
+
+Recently, the OptaPlanner team is shifting towards cloud-native applications.
+As a part of this effort, we migrated the OptaWeb Employee Rostering Backend Application from WildFly to Spring Boot.
+
+OptaWeb Employee Rostering is a web application that optimizes shift scheduling for maximum efficiency and fairness for employees.
+In this post we’ll go over how the OptaPlanner web application was moved to the Spring framework, and the key features that were added on top of the old backend.
+
+== The Problem
+
+Let’s start with the problem we’re trying to solve.
+Here is a visual overview of the https://www.optaplanner.org/learn/useCases/employeeRostering.html[Employee Rostering problem]:
+
+image::employeeRosteringProblem.png[]
+
+Many company workplaces need solutions to this problem, from hospitals to call centers.
+There are many factors to consider when assigning shifts to employees, such as required skills, employee availabilities, and employee fairness.
+So how do we find the best solution?
+With OptaPlanner, of course!
+
+== Domain Model
+
+Before we dive into the code, we first need a domain model.
+A domain model represents the entities (Java objects) and how they relate with other entities in the application.
+It tells us the behaviour of the objects and how they change during solving.
+Here's a class diagram of the Employee Rostering model, where employees are nurses:
+
+image::employeeRosteringClassDiagram.png[]
+
+Each of the blocks above represents a Java entity object in our application, which will be saved in a Spring JPA Database for solving.
+The arrows show how certain entities map to others, so changing one during the solving process may affect another.
+The `*` and `1` symbols on the arrows represent a many-to-one mapping of the entities in the database.
+
+Here's a simple example of the `Skill` entity:
+
+[source,java]
+----
+@Entity
+@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
+public class Skill extends AbstractPersistable {
```

Comment:
I'd consider if we want to expose the internals, such as `AbstractPersistable`. Without explaining what it is, it may be confusing to a first time reader. And if you explain it, the blog will be too detailed. 

If we call this a simplified entity, perhaps we can get away with removing these details?